flink-ml-lib/src/main/java/org/apache/flink/ml/feature/minmaxscaler/MinMaxScalerModel.java

```diff
@@ -0,0 +1,181 @@
+                                            broadcastModelKey,
+                                            getMax(),
+                                            getMin(),
+                                            getFeaturesCol()),
+                                    outputTypeInfo);
+                        });
+        return new Table[] {tEnv.fromDataStream(output)};
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+        ReadWriteUtils.saveModelData(
+                MinMaxScalerModelData.getModelDataStream(modelDataTable),
+                path,
+                new MinMaxScalerModelData.ModelDataEncoder());
+    }
+
+    /**
+     * Loads model data from path.
+     *
+     * @param env Stream execution environment.
+     * @param path Model path.
+     * @return MinMaxScalerModel model.
+     */
+    public static MinMaxScalerModel load(StreamExecutionEnvironment env, String path)
+            throws IOException {
+        StreamTableEnvironment tEnv = StreamTableEnvironment.create(env);
+        MinMaxScalerModel model = ReadWriteUtils.loadStageParam(path);
+        DataStream<MinMaxScalerModelData> modelData =
+                ReadWriteUtils.loadModelData(
+                        env, path, new MinMaxScalerModelData.ModelDataDecoder());
+        return model.setModelData(tEnv.fromDataStream(modelData));
+    }
+
+    /** This operator loads model data and predicts result. */
+    private static class PredictOutputFunction extends RichMapFunction<Row, Row> {
+        private final String featureCol;
+        private MinMaxScalerModelData minMaxScalerModelData;
+        private final double upperBound;
+        private final double lowerBound;
+        private final String broadcastKey;
+        private DenseVector maxVector;
+        private DenseVector minVector;
+
+        public PredictOutputFunction(
+                String broadcastKey, double upperBound, double lowerBound, String featureCol) {
+            this.upperBound = upperBound;
+            this.lowerBound = lowerBound;
+            this.broadcastKey = broadcastKey;
+            this.featureCol = featureCol;
+        }
+
+        @Override
+        public Row map(Row row) {
+            if (minMaxScalerModelData == null) {
+                minMaxScalerModelData =
+                        (MinMaxScalerModelData)
+                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
+                maxVector = minMaxScalerModelData.maxVector;
+                minVector = minMaxScalerModelData.minVector;
+            }
+            DenseVector feature = (DenseVector) row.getField(featureCol);
+            DenseVector outputVector = new DenseVector(maxVector.size());
+            if (feature != null) {
+                for (int i = 0; i < maxVector.size(); ++i) {
```

Comment:
This for loop might also be replaced by BLAS operations.

Reply:
There is a judge about max-min==0 exists in the loop, then we can not use blas here. 
If we want to use blas, we need another loop to prepare data. I think it is more expensive.

Reply:
Just for this code, I think there is a simple way to do the max == min judgement just once, like follows.
```java

        public Row map(Row row) {
            if (minMaxScalerModelData == null) {
                minMaxScalerModelData =
                        (MinMaxScalerModelData)
                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
                maxVector = minMaxScalerModelData.maxVector;
                minVector = minMaxScalerModelData.minVector;
                for (int i = 0; i < maxVector.size(); ++i) {
                    if ((minVector.values[i] - maxVector.values[i]) == 0.0) {
                        maxVector.values[i] = maxVector.values[i] + 1.0;
                        minVector.values[i] = minVector.values[i] - 1.0;
                    }
                }
            }
            DenseVector feature = (DenseVector) row.getField(featureCol);
            DenseVector outputVector = new DenseVector(maxVector.size());
            if (feature != null) {
                for (int i = 0; i < maxVector.size(); ++i) {
                    outputVector.values[i] =
                            (feature.values[i] - minVector.values[i])
                                    / (maxVector.values[i] - minVector.values[i])
                                    * (upperBound - lowerBound)
                                    + lowerBound;
                }
                return Row.join(row, Row.of(outputVector));
            } else {
                throw new RuntimeException("Feature value is null, please check your input data.");
            }
        }
```
With this code, now it is possible to apply BLAS operations.

I noticed that Spark also uses for loop in this and the situation above, so I agree that we may postpone any further optimizations like this after this PR.

Reply:
If featureValue[i] != maxVector[i], above code may not get the same result as original code.
In read world, data for train and predict are different, featureValue[i] != maxVector[i] is existed.

Reply:
> if the feature vector exists, but its dimension is different from that of maxVector, or its max/min value exceeds maxVector/minVector's range, then maybe we should also throw exception.

If `featureValue[i] != maxVector[i]`, the input data would have violated the rule above and would throw exceptions. So there would be no result generated in both cases, right?

From this, I think we should also add tests about illegal inputs, like `OneHotEncoderTest.testNonIndexedPredictData`.

Reply:
I don‘t think so. 
If featureValue[i] != maxVector[i],  predict data may have mo problem. For train data defines the rule and we can't constrain predict data follow this rule. 

For example, in train data one feature is people's age, all feature values in the train data is 20 years old. If predict data has a sample which feature is 21 years old which will be scaled to 0.5*(max - min).

Reply:
Got it. Then how about this one?
```java
        @Override
        public Row map(Row row) {
            if (minMaxScalerModelData == null) {
                minMaxScalerModelData =
                        (MinMaxScalerModelData)
                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
                maxVector = minMaxScalerModelData.maxVector;
                minVector = minMaxScalerModelData.minVector;
                scaleVector = new DenseVector(minVector.size());
                offsetVector = new DenseVector(minVector.size());
                for (int i = 0; i < maxVector.size(); ++i) {
                    if ((minVector.values[i] - maxVector.values[i]) == 0.0) {
                        scaleVector.values[i] = Double.POSITIVE_INFINITY;
                        offsetVector.values[i] = 0.5;
                    } else {
                        scaleVector.values[i] = maxVector.values[i] - minVector.values[i];
                        offsetVector.values[i] = 0.0;
                    }
                }
            }
            DenseVector feature = (DenseVector) row.getField(featureCol);
            DenseVector outputVector = new DenseVector(maxVector.size());
            if (feature != null) {
                for (int i = 0; i < maxVector.size(); ++i) {
                    outputVector.values[i] =
                            ((feature.values[i] - minVector.values[i])
                                                    / scaleVector.values[i]
                                                    + offsetVector.values[i])
                                            * (upperBound - lowerBound)
                                    + lowerBound;
                }
                return Row.join(row, Row.of(outputVector));
            } else {
                throw new RuntimeException("Feature value is null, please check your input data.");
            }
        }
```

This code should meet the requirements without using if conditions.

Reply:
Besides, could you please add a test case about maxVector[i] == minVector[i] && featureValue[i] != maxVector[i]? I think it would be helpful to test these corner cases.

Reply:
Even better:
```java

        @Override
        public Row map(Row row) {
            if (minMaxScalerModelData == null) {
                minMaxScalerModelData =
                        (MinMaxScalerModelData)
                                getRuntimeContext().getBroadcastVariable(broadcastKey).get(0);
                maxVector = minMaxScalerModelData.maxVector;
                minVector = minMaxScalerModelData.minVector;
                scaleVector = new DenseVector(minVector.size());
                offsetVector = new DenseVector(minVector.size());
                for (int i = 0; i < maxVector.size(); ++i) {
                    if ((minVector.values[i] - maxVector.values[i]) == 0.0) {
                        scaleVector.values[i] = 0.0;
                        offsetVector.values[i] = (upperBound + lowerBound) / 2;
                    } else {
                        scaleVector.values[i] = (upperBound - lowerBound) / (maxVector.values[i] - minVector.values[i]);
                        offsetVector.values[i] = lowerBound - minVector.values[i] * scaleVector.values[i];
                    }
                }
            }
            DenseVector feature = (DenseVector) row.getField(featureCol);
            DenseVector outputVector = new DenseVector(maxVector.size());
            if (feature != null) {
                for (int i = 0; i < maxVector.size(); ++i) {
                    outputVector.values[i] = feature.values[i] * scaleVector.values[i] + offsetVector.values[i];
                }
                return Row.join(row, Row.of(outputVector));
            } else {
                throw new RuntimeException("Feature value is null, please check your input data.");
            }
        }
```

With this implementation, there would be only one multiplication and one addition in the for loop, which can definitely be replaced by `BLAS.axpy`. It would cause some precision loss (less then 1e-16 on this PR's test data), but I think the performance improvement worth it.

Reply:
I got it. But in BLAS.axpy, a is a double value which is a constant, not a value in vector.
If use BLAS, we need two BLAS func: vec.*vec and axpy. This means we need two vector loop.  I don't think it's more efficiency. 

Your code can avoid the judge of "if ((minVector.values[i] - maxVector.values[i]) == 0.0)" in transform, I will refine my code as you done.